src/charm.py

```diff
@@ -309,20 +321,30 @@ def _validate(self):
             if not any(service == item.value for item in ValidServiceTypes):
                 raise ValueError(f"error in services config: invalid service {service!r}")
 
-        # Validate relations.
+        # Validate admin relation.
         self.database_connections()
         if not self._state.schema_ready:
             raise ValueError("admin:temporal relation: schema is not ready")
 
+        # Validate OpenFGA relation.
         if self.config["auth-enabled"]:
             if not self._state.openfga:
                 raise ValueError("openfga:temporal relation not ready")
-            missing_params = self._check_missing_openfga_params()
+            missing_params = self._check_missing_params(self._state.openfga, REQUIRED_OPENFGA_KEYS)
             if len(missing_params) > 0:
                 raise ValueError(f"openfga:missing parameters {missing_params!r}")
             if not self._state.openfga["auth_model_id"]:
                 raise ValueError("missing openfga authorization model")
 
+        # Validate S3 relation.
+        if self._state.s3:
+            missing_params = self._check_missing_params(self._state.s3, REQUIRED_S3_PARAMETERS)
+            if len(missing_params) > 0:
+                raise ValueError(f"s3:missing parameters {missing_params!r}")
+
+            if not self._state.s3["bucket_created"]:
```

Comment:
Is the "bucket_created" key guaranteed to be in the s3 dict?

Reply:
It is, but i've changed this to a `.get()` instead just to be safe.